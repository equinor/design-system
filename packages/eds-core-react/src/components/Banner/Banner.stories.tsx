import { useState, useEffect } from 'react'
import { ComponentMeta, Story } from '@storybook/react'
import { Banner, Icon, Button, BannerProps, EdsProvider, Density } from '../..'
import { save, thumbs_up, thumbs_down, mood_sad } from '@equinor/eds-icons'
import page from './Banner.docs.mdx'

const icons = {
  save,
  thumbs_up,
  thumbs_down,
  mood_sad,
}

Icon.add(icons)

export default {
  title: 'Feedback/Banner',
  component: Banner,
  subcomponents: {
    Message: Banner.Message,
    Icon: Banner.Icon,
    Actions: Banner.Actions,
  },
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Banner>

export const Introduction: Story<BannerProps> = () => {
  //  Note: This example is not interactive, as Storybook
  // doesn't yet support to manipulate subcomponents via Storybook Args
  return (
    <Banner>
      <Banner.Message>
        This tag is not being preserved yet. Click start preservation to enable
        writing preservation records.
      </Banner.Message>
    </Banner>
  )
}

export const TextAndIcon: Story<BannerProps> = () => (
  <>
    <Banner>
      <Banner.Icon>
        <Icon name="thumbs_up" />
      </Banner.Icon>
      <Banner.Message>
        We are in the making of a new design for this page.
      </Banner.Message>
    </Banner>
    <Banner>
      <Banner.Icon variant="warning">
        <Icon name="thumbs_down" />
      </Banner.Icon>
      <Banner.Message>Some warning information.</Banner.Message>
    </Banner>
  </>
)
TextAndIcon.storyName = 'Text and icon'

export const TextAndAction: Story<BannerProps> = () => (
  <>
    <Banner>
      <Banner.Message>
        You are signed on with another account than Equinor account:
        name.lastname@mail.com.
      </Banner.Message>
      <Banner.Actions>
        <Button variant="ghost">Sign out</Button>
        <Button variant="ghost">OK</Button>
      </Banner.Actions>
    </Banner>
    <Banner>
      <Banner.Message>
        You are signed on with another account than Equinor account:
        name.lastname@mail.com. This means you have to do something to be able
        to use this service.
      </Banner.Message>
      <Banner.Actions placement="bottom">
        <Button variant="ghost">Sign out</Button>
        <Button variant="ghost">OK</Button>
      </Banner.Actions>
    </Banner>
  </>
)
TextAndAction.storyName = 'Text and action'

export const TextAndIconAndAction: Story<BannerProps> = () => (
  <>
    <Banner>
      <Banner.Icon variant="warning">
        <Icon name="mood_sad" />
      </Banner.Icon>
      <Banner.Message>
        This tag is not being preserved yet. Click start preservation to enable
        writing preservation records.
      </Banner.Message>
      <Banner.Actions>
        <Button>Action</Button>
      </Banner.Actions>
    </Banner>
    <Banner>
      <Banner.Icon>
        <Icon name="save" />
      </Banner.Icon>
      <Banner.Message>
        I&apos;m such a really really long message about some sad saving news
        that there is not enough space for the actions on my left. That&apos;s
        why the actions have been located at the bottom using the placement prop
        instead.
      </Banner.Message>
      <Banner.Actions placement="bottom">
        <Button>First action</Button>
        <Button color="secondary">Second action</Button>
      </Banner.Actions>
    </Banner>
  </>
)
TextAndIconAndAction.storyName = 'Text and icon and actions'

export const Compact: Story<BannerProps> = () => {
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  return (
    <EdsProvider density={density}>
      <Banner>
        <Banner.Icon variant="warning">
          <Icon name="mood_sad" />
        </Banner.Icon>
        <Banner.Message>
          This tag is not being preserved yet. Click start preservation to
          enable writing preservation records.
        </Banner.Message>
        <Banner.Actions>
          <Button>Action</Button>
        </Banner.Actions>
      </Banner>
      <Banner>
        <Banner.Icon>
          <Icon name="save" />
        </Banner.Icon>
        <Banner.Message>
          I&apos;m such a really really long message about some sad saving news
          that there is not enough space for the actions on my left. That&apos;s
          why the actions have been located at the bottom using the placement
          prop instead.
        </Banner.Message>
        <Banner.Actions placement="bottom">
          <Button>First action</Button>
          <Button color="secondary">Second action</Button>
        </Banner.Actions>
      </Banner>
    </EdsProvider>
  )
}
