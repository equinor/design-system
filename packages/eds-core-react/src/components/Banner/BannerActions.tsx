import { forwardRef, HTMLAttributes, ReactNode } from 'react'
import styled, { css } from 'styled-components'

type BannerActionsPlacement = 'bottom' | 'left'

type StyledBannerActionsProps = {
  /** Where to place the actions */
  $placement: BannerActionsPlacement
}

const StyledBannerActions = styled.div<StyledBannerActionsProps>(
  ({ theme, $placement }) => {
    return css`
      display: flex;
      margin-left: ${theme.spacings.left};
      grid-gap: 8px;
      grid-column: ${$placement === 'bottom' ? '1/-1' : 'auto'};
      ${$placement === 'bottom' && {
        marginTop: theme.spacings.top,
        marginLeft: '0',
      }}
    `
  },
)

export type BannerActionsProps = {
  children: ReactNode
  placement?: BannerActionsPlacement
} & HTMLAttributes<HTMLDivElement>

export const BannerActions = forwardRef<HTMLDivElement, BannerActionsProps>(
  function BannerActions({ children, placement = 'left', ...rest }, ref) {
    const props = {
      ref,
      ...rest,
    }
    return (
      <StyledBannerActions $placement={placement} {...props}>
        {children}
      </StyledBannerActions>
    )
  },
)
