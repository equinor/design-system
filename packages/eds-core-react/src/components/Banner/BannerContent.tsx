import { forwardRef, ReactNode } from 'react'

export type BannerContentProps = {
  /** Complex content with any HTML elements */
  children: ReactNode
} & React.HTMLAttributes<HTMLDivElement>

export const BannerContent = forwardRef<HTMLDivElement, BannerContentProps>(
  function BannerContent({ children, ...rest }, ref) {
    return (
      <div ref={ref} {...rest}>
        {children}
      </div>
    )
  },
)
