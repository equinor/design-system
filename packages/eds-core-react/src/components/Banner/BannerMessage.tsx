import { forwardRef } from 'react'
import styled from 'styled-components'
import { Typography } from '../Typography'
import { TypographyProps } from '../Typography/Typography'

const StyledBannerMessage = styled(Typography)``

export type BannerMessageProps = {
  /** Text content */
  children: string
} & Omit<TypographyProps, 'children'>

export const BannerMessage = forwardRef<HTMLElement, BannerMessageProps>(
  function BannerMessage({ children, ...rest }, ref) {
    const props = {
      ref,
      ...rest,
    }

    return (
      <StyledBannerMessage variant="body_long" {...props}>
        {children}
      </StyledBannerMessage>
    )
  },
)
