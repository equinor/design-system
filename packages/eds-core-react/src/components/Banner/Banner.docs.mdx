import { Links, PropsTable, Story } from './../../../.storybook/components'
import { Banner } from '.'

# Banner

Displays important notifications and related optional actions.

<Links 
    ariaUrl="https://www.w3.org/TR/2017/REC-wai-aria-1.1-20171214/#banner"
    figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17611%3A0"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/245b05-banner/b/22c859"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Banner/Banner.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Story id="feedback-banner--introduction" />
<PropsTable />

## Usage

A banner displays an important, brief message, and provides actions for users to address. 
It requires a user action to be dismissed.

<ul>
    <li>Banners are interruptive, but their level of interruption should match 
        the information they hold and the context in which they appear.</li>
    <li>Only one banner should be shown at a time.</li>
</ul>

```tsx
import { Banner } from '@equinor/eds-core-react'

<Banner>
  <Banner.Message>
    Banner message comes here.
  </Banner.Message>
</Banner>
```

## Examples

### Text and icon

Banners can supplement their message using a supporting icon.

<Story id="feedback-banner--text-and-icon" />

### Text and action

<ul>
  <li>Buttons in banners should directly relate to a banner’s message and clearly represent the banner’s action.</li>
  <li>Buttons must be labelled with text, not icons, for clarity.</li>
  <li>Place buttons under the banner message, or on the same line if there is enough room to fit both.</li>
</ul>

<Story id="feedback-banner--text-and-action" />

### Text and icon and action

<ul>
  <li>The banner message communicates a change or error within an app.</li>
  <li>Banners should be considered as part of your overall in-app messaging strategy.</li>
</ul>

<Story id="feedback-banner--text-and-icon-and-action" />

### Compact

Compact `Banner` using `EdsProvider`.

<Story id="feedback-banner--compact" />
