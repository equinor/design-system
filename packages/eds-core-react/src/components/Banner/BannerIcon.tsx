import {
  forwardRef,
  HTMLAttributes,
  ReactNode,
  Children as ReactChildren,
  isValidElement,
  cloneElement,
} from 'react'
import styled, { css } from 'styled-components'
import * as tokens from './Banner.tokens'
import { Icon } from '../Icon'
import { bordersTemplate } from '@equinor/eds-utils'

type BannerIconVariant = 'info' | 'warning'

type StyledBannerIconProps = {
  variant: BannerIconVariant
}

const { info, warning } = tokens

const StyledBannerIcon = styled.span<StyledBannerIconProps>(
  ({ theme, variant }) => {
    return css`
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      ${bordersTemplate(theme.entities.icon.border)};
      background-color: ${variant === 'warning'
        ? warning.entities.icon.background
        : info.entities.icon.background};
      width: ${theme.entities.icon.width};
      height: ${theme.entities.icon.height};
      margin-right: ${theme.spacings.right};
    `
  },
)

export type BannerIconProps = {
  /** Which icon background and fill color to use. Info = green, warning = red */
  variant?: BannerIconVariant
  /** @ignore */
  children: ReactNode
} & HTMLAttributes<HTMLSpanElement>

export const BannerIcon = forwardRef<HTMLSpanElement, BannerIconProps>(
  function BannerIcon({ children, variant = 'info', ...rest }, ref) {
    const childrenWithColor = ReactChildren.map(children, (child) => {
      const color =
        variant === 'warning'
          ? warning.entities.icon.typography.color
          : info.entities.icon.typography.color
      return (
        (isValidElement(child) &&
          child.type === Icon &&
          cloneElement(child, {
            color,
          })) ||
        child
      )
    })

    const props = {
      ref,
      ...rest,
    }

    return (
      <StyledBannerIcon variant={variant} {...props}>
        {childrenWithColor}
      </StyledBannerIcon>
    )
  },
)
