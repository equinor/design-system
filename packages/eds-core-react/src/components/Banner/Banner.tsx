import {
  forwardRef,
  HTMLAttributes,
  isValidElement,
  ReactNode,
  Children as ReactChildren,
} from 'react'
import styled, { css, ThemeProvider } from 'styled-components'
import { spacingsTemplate, useToken } from '@equinor/eds-utils'
import { Paper } from '../Paper'
import type { Elevations } from '@equinor/eds-tokens'
import { enabled as bannerToken } from './Banner.tokens'
import { Divider } from '../Divider'
import { BannerIcon } from './BannerIcon'
import { useEds } from '../EdsProvider'

type AvailableElevations = keyof Pick<Elevations, 'none' | 'raised' | 'overlay'>

type ContentProps = {
  hasIcon: boolean
}

const Content = styled.div<ContentProps>(({ theme, hasIcon }) => {
  return css`
    ${spacingsTemplate(theme.spacings)}

    display: grid;
    grid-template-columns: ${hasIcon ? 'min-content 1fr auto' : '1fr auto'};
    align-items: center;
    background-color: ${theme.background};
  `
})

const NonMarginDivider = styled(Divider)`
  margin: 0;
  height: 2px;
`

export type BannerProps = {
  elevation?: AvailableElevations
  children: ReactNode
} & HTMLAttributes<HTMLDivElement>

export const Banner = forwardRef<HTMLDivElement, BannerProps>(function Banner(
  { children, className, elevation = 'none', ...rest },
  ref,
) {
  const childrenWhereBannerIcon: boolean[] = ReactChildren.map(
    children,
    (child) => {
      return isValidElement(child) && child.type === BannerIcon
    },
  )
  const hasIcon = childrenWhereBannerIcon.some((bool) => bool)

  const props = {
    ref,
    ...rest,
  }

  const { density } = useEds()
  const token = useToken({ density }, bannerToken)

  return (
    <ThemeProvider theme={token}>
      <Paper
        {...props}
        className={className}
        elevation={elevation}
        role="alert"
      >
        <Content hasIcon={hasIcon}>{children}</Content>
        <NonMarginDivider color="light" />
      </Paper>
    </ThemeProvider>
  )
})
