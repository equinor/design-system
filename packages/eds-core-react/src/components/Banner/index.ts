import { Banner as BaseBanner, BannerProps } from './Banner'
import { BannerIcon, BannerIconProps } from './BannerIcon'
import { BannerMessage, BannerMessageProps } from './BannerMessage'
import { BannerActions, BannerActionsProps } from './BannerActions'
import { BannerContent, BannerContentProps } from './BannerContent'

type BannerCompoundProps = typeof BaseBanner & {
  Icon: typeof BannerIcon
  Message: typeof BannerMessage
  Actions: typeof BannerActions
  Content: typeof BannerContent
}

const Banner = BaseBanner as BannerCompoundProps
Banner.Icon = BannerIcon
Banner.Message = BannerMessage
Banner.Actions = BannerActions
Banner.Content = BannerContent

Banner.Icon.displayName = 'Banner.Icon'
Banner.Message.displayName = 'Banner.Message'
Banner.Actions.displayName = 'Banner.Actions'
Banner.Content.displayName = 'Banner.Content'

export { Banner, BannerIcon, BannerMessage, BannerActions, BannerContent }
export type {
  BannerProps,
  BannerMessageProps,
  BannerIconProps,
  BannerActionsProps,
  BannerContentProps,
}
