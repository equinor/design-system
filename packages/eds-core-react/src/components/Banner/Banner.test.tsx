/* eslint-disable no-undef */
import { render, cleanup, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { add } from '@equinor/eds-icons'
import { Banner } from '.'
import { Icon } from '../Icon'
import * as tokens from './Banner.tokens'

Icon.add({ add })

const { info, warning } = tokens

const StyledBanner = styled(Banner)`
  position: relative;
`

const rgbaTrim = (x: string) => x.split(' ').join('')

afterEach(cleanup)

describe('Banner', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <Banner>
        <Banner.Icon>
          <Icon data={add} />
        </Banner.Icon>
        <Banner.Message>Banner message</Banner.Message>
      </Banner>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test when only text', async () => {
    const { container } = render(
      <StyledBanner>
        <Banner.Message>Banner message</Banner.Message>
      </StyledBanner>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test when text and icon', async () => {
    const { container } = render(
      <StyledBanner>
        <Banner.Icon>
          <Icon name="add" />
        </Banner.Icon>
        <Banner.Message>Banner message</Banner.Message>
      </StyledBanner>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test when text and actions', async () => {
    const actionButtonText = 'Banner action button text'
    const { container } = render(
      <StyledBanner>
        <Banner.Message>Banner message</Banner.Message>
        <Banner.Actions>
          <button type="button">{actionButtonText}</button>
        </Banner.Actions>
      </StyledBanner>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Can extend the css for the component', () => {
    render(
      <StyledBanner>
        <Banner.Message>styled banner</Banner.Message>
      </StyledBanner>,
    )
    expect(screen.getByRole('alert')).toHaveStyleRule('position', 'relative')
  })
  it('Has provided Message', () => {
    const bannerText = 'Banner test message'
    render(
      <Banner>
        <Banner.Message>{bannerText}</Banner.Message>
      </Banner>,
    )
    expect(screen.getByText(bannerText)).toBeDefined()
  })
  it('Has provided Icon', () => {
    const bannerText = 'Banner test'
    const iconTestId = 'banner-icon-test'
    render(
      <Banner>
        <Banner.Icon>
          <Icon name="add" data-testid={iconTestId} />
        </Banner.Icon>
        <Banner.Message>{bannerText}</Banner.Message>
      </Banner>,
    )
    expect(screen.getByText(bannerText)).toBeDefined()
    expect(screen.getByTestId(iconTestId)).toBeDefined()
  })
  it('Has provided Actions', () => {
    const bannerText = 'Banner test'
    const actionButtonText = 'Banner action button text'
    render(
      <Banner>
        <Banner.Message>{bannerText}</Banner.Message>
        <Banner.Actions>
          <button type="button">{actionButtonText}</button>
        </Banner.Actions>
      </Banner>,
    )
    expect(screen.getByText(bannerText)).toBeDefined()
    expect(screen.getByText(actionButtonText)).toBeDefined()
  })

  it('Has correct default icon styling as info color', () => {
    const bannerText = 'Banner test'
    const iconWrapperTestId = 'banner-icon-wrapper-test'
    render(
      <Banner>
        <Banner.Icon data-testid={iconWrapperTestId}>
          <Icon name="add" data-testid="icon" />
        </Banner.Icon>
        <Banner.Message>{bannerText}</Banner.Message>
      </Banner>,
    )
    const iconSvg = screen.getByTestId('icon')
    expect(screen.queryByTestId(iconWrapperTestId)).toHaveStyleRule(
      'background-color',
      rgbaTrim(info.entities.icon.background),
    )
    expect(iconSvg).toHaveAttribute('fill', info.entities.icon.typography.color)
  })
  it('Has correct warning icon styling', () => {
    const bannerText = 'Banner test'
    const iconWrapperTestId = 'banner-icon-wrapper-test'
    render(
      <Banner>
        <Banner.Icon variant="warning" data-testid={iconWrapperTestId}>
          <Icon name="add" data-testid="icon" />
        </Banner.Icon>
        <Banner.Message>{bannerText}</Banner.Message>
      </Banner>,
    )
    const iconSvg = screen.getByTestId('icon')

    expect(screen.queryByTestId(iconWrapperTestId)).toHaveStyleRule(
      'background-color',
      rgbaTrim(warning.entities.icon.background),
    )
    expect(iconSvg).toHaveAttribute(
      'fill',
      warning.entities.icon.typography.color,
    )
  })
})
