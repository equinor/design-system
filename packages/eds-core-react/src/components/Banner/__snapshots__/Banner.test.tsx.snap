// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  background: var(--eds_ui_background__default,rgba(255,255,255,1));
  box-shadow: 0 0 1px rgba(0,0,0,0.14);
}

.c4 {
  border: none;
  background-color: var(--eds_ui_background__light,rgba(247,247,247,1));
  margin-top: 16px;
  margin-bottom: calc(16px - 1px);
  height: 1px;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 100px;
  background-color: var(--eds_infographic_primary__moss_green_13,rgba(222,237,238,1));
  width: 40px;
  height: 40px;
  margin-right: 16px;
}

.c1 {
  padding-left: 16px;
  padding-top: 16px;
  padding-right: 16px;
  padding-bottom: 16px;
  display: grid;
  grid-template-columns: min-content 1fr auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--eds_ui_background__default,rgba(255,255,255,1));
}

.c5 {
  margin: 0;
  height: 2px;
}

.c3 {
  margin: 0;
  color: var(--eds_paragraph__body_long_color,rgba(61,61,61,1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.500em;
  text-align: left;
}

<div
    class="c0"
    role="alert"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="var(--eds_infographic_primary__moss_green_100, rgba(0, 112, 121, 1))"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            clip-rule="evenodd"
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2Z"
            data-testid="eds-icon-path"
            fill-rule="evenodd"
            height="24"
          />
        </svg>
      </span>
      <p
        class="c3 "
      >
        Banner message
      </p>
    </div>
    <hr
      class="c4 c5"
    />
  </div>
</DocumentFragment>
`;
